mash_client/cli_utils.py

```diff
@@ -89,11 +89,11 @@ def get_config(cli_context):
     Use ChainMap to build config values based on
     command line args, config and defaults.
     """
-    config_path = cli_context['config'] or default_config
+    config_path = cli_context['config_dir'] or default_config_dir
 
     config_values = {}
     with suppress(Exception):
-        with open(config_path) as config_file:
+        with open(config_path + 'config.yaml') as config_file:
```

Comment:
If we support a command line option to point to another config file, such as `-f my-config` then the name of the config should not be hard coded to be `config.yaml`. If it is not intended to allow the user to specify the location of the config file, which I think it should be, then I do not understand whay the change is necessary.

Reply:
The context of the change is not easy to see since the entire CLI changes now to handle auth. Because there will be multiple config files (new auth tokens file) I think the better option is for the user to provide a config directory rather than the entire config file path. The client then uses the dir to know where to look to find config file and to save/retrieve tokens.

The command line option is now config_dir so this variable would be better named config_dir instead of config_path.

Reply:
Thanks for the clarification and update. I am thinking that we might not want to store app managed content, i.e. the tokens next to a file that is user modifiable. It is probably good to keep those separated and track a "tocken_dir" that is hard coded, and then should point to "/var/run/user/UID/mash-client". This way we make it explicitly "obvious" about what the user can touch and should not touch. Of course the token file name would need to be prefixed by some identifier in case the user has multiple mash server they connect to.

Reply:
I don't think it needs to be this complicated:

- If the client is storing info under /var... then it needs special permissions which it shouldn't require since it's a user app.
- If a user chooses to muck with their tokens file then they would login again to fix things. (This is very simple to fix)
- The precedence I see with other command line apps is to store tokens alongside config in files such as tokens.json, access_tokens.db etc etc. What makes mash client so special that it can't do the same?
- From a user perspective I prefer having all application artifacts in one location and this seems to match the precedence.

As far as handling multiple sessions I think we should start with YAGNI. If down the road this is actually a need then we should consider it but for now there's no use case for multiple production mash servers.

Reply:
No you do not need special permissions to write to `/var/run/user/$UID` this is setup by systemd and the user has permission as you can see with 'ls -ld /var/run/user/$UID/*` and you will see that there are a number of processes that write data there that are executed with user permission.

If we follow YAGNI we do not need to make this configurable at all. Then let's remove the config_dir switch and put everything in `~/.config/mash-client` that is the most simple approach.

As far as existing solutions is concerned based on a sample size of 2 ;) it's 50-50. osc stores data separately `.osc_cookiejar`, the again osc scribbles everything to one's home directory which could be considered storing things side by side, and az tools stores things side by side in a specific hidden directory.

So following your arguments, let's drop the command line option and write everything to `~/.config`

Reply:
> This way we make it explicitly "obvious" about what the user can touch and should not touch.

The advantage to this route of "hiding" the tokens file is to protect the user from themselves? I think we should give the user the benefit of the doubt. Given a mucked up tokens file is extremely simple to fix I think it's perfectly fine to leave in the config directory.

Anyway, I think this discussion is generally opinionated and we should stick with the original implementation. Unless there is a functional problem that exists.